src/adapters/store.js

```diff
@@ -3,111 +3,169 @@ import Error from '../adapters/error'
 import {version} from '../../config/constants.yml'
 import ExtendedString from "../libs/string";
 
-let moduleConfig = nconf.get().store
+export default class Store {
 
-const AbStore = require(`../libs/${moduleConfig.name}`)
-const abstractStore = new AbStore(moduleConfig.endpoint)
+  constructor() {
+    // get store from window if already initialized
+    if (window.__store) {
+      return window.__store
+    }
+    // if store not initialized, use this
+    window.__store = this
 
-function Store() {
-  this.isRegisterd = false
-  if(!window.__existingStore) {
-    window.__existingStore = true
-    listen('store_poi', (poi) => {
-      this.add(poi)
-    })
-    listen('del_poi', (poi) => {
-      this.del(poi)
-    })
-    listen('store_center', (loc) => {
-      this.setLastLocation(loc)
-    })
-    listen('store_clear', () => {
-      this.clear()
-    })
+    // init stores
+    const LocalStore = require(`../libs/local_store`)
+    this.localStore = new LocalStore()
+    this.loggedIn = false
+    this.abstractStore = this.localStore
+    const MasqStore = require(`../libs/masq`)
+    this.masqConfig = nconf.get().masq
+    this.masqStore = new MasqStore(this.masqConfig)
+    const masqInitPromise = this.masqStore.onConnect()
+    const masqInitialized = false
+
+    // use abstract store for each operation that
+    // should use masqStore when logged in and localStore when not logged in
+
+    return this
   }
-}
 
-Store.prototype.getAllPois = async function() {
-  try {
-    return await abstractStore.getAllPois()
-  } catch (e) {
-    Error.sendOnce('store', 'getAllPois', 'error getting pois', e)
-    throw e
+  async checkInit(target, name, descriptor) {
+    if (this.loggedIn && !this.masqInitialized) {
+      const alreadyLoggedIntoMasq = await this.masqStore.isLoggedIn()
+      if (alreadyLoggedIntoMasq) {
+        this.abstractStore = this.masqStore
+      }
+      await this.masqInitPromise
+      this.masqInitialized = true
+    }
   }
-}
 
-Store.prototype.getLastLocation = async function() {
-  try {
-    return await abstractStore.get(`qmaps_v${version}_last_location`)
-  } catch (e) {
-    Error.sendOnce('store', 'getLastLocation', 'error getting location', e)
-    return null
+  async login() {
+    await this.checkInit()
+    try {
+      const loginParams = {
+        endpoint: this.masqConfig.endpoint,
+        url: window.location.origin + window.location.pathname,
+        title: this.masqConfig.title,
+        desc: this.masqConfig.desc,
+        icon: this.masqConfig.icon
+      }
+
+      await this.masqStore.login(loginParams)
+    } catch (e) {
+      Error.sendOnce('store', 'login', 'error logging in', e)
+      throw e
```

Comment:
this error is catched twice (here and masqLoginPanel)

Reply:
removed error logging in panel/login_masq